### PR TITLE
feat(upload): Add batch actions for creation and invitation

### PR DESCRIPTION
**Pull request by Michaelvilleneuve**
Ajout d'actions par lot sur la page d'upload d'usagers. 
Pour les utiliser il suffit de sélectionner les utilisateurs sur lesquels effectuer une action en utilisant la nouvelle barre de sélection : 
image

Puis d'utiliser le nouveau dropdown qui apparait : 
image

À savoir, comme évoqué avec Salma, déclencher des invitations par batch sur des usagers sans compte déclenche également au passage la création de compte pour ceux-ci avant d'effectuer l'action souhaitée. 

Issue :  https://github.com/betagouv/rdv-insertion/issues/484

### CODE REVIEW

File: app/javascript/lib/retrieveRelevantOrganisation.js [hunk 1]
```diff
@@ -5,19 +5,22 @@ import searchOrganisations from "../react/actions/searchOrganisations";
...
+  return retrieveThroughGeolocalisation(departmentNumber, applicantFullAddress, options);
 };
 
-const retrieveThroughSearchTerms = async (departmentNumber, organisationSearchTerms) => {
+const retrieveThroughSearchTerms = async (departmentNumber, organisationSearchTerms, options = { raiseError: true }) => {
   const result = await searchOrganisations(departmentNumber, organisationSearchTerms);
   if (result.success && result.matching_organisations.length === 1) {
     return result.matching_organisations[0];
+  } if (options.raiseError === false) {
+    return null;
```
Comment (aminedhobb) [inline app/javascript/lib/retrieveRelevantOrganisation.js:23]: ça ne marche pas quand j'essaie en local, la modale apparait chez moi: 
image

File: app/javascript/react/pages/ApplicantsUpload.jsx [hunk 2]
```diff
@@ -131,15 +132,13 @@ export default function ApplicantsUpload({
               columnNames,
               currentAgent
             );
-            applicantsFromList.push(applicant);
+            applicants.addApplicant(applicant);
```
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/react/pages/ApplicantsUpload.jsx:135]: D'ailleurs la liste reste si on fait "précédent - suivant" sur son navigateur. Je sais pas si c'est ce qu'on veut 🤷‍♂️

File: app/javascript/react/models/Applicant.js [hunk 3]
```diff
@@ -103,7 +131,70 @@ export default class Applicant {
     return role;
   }
 
+  async inviteBy(format, isDepartmentLevel) {
+    if (!this.createdAt) await this.createAccount()
```
Comment (aminedhobb) [inline app/javascript/react/models/Applicant.js:135]: Je pense qu'il faut rajouter une option `raiseError` ici aussi: 

image

**Review by aminedhobb — COMMENTED** [review]
Merci @Michaelvilleneuve c'est vraiment génial, notre app front gagne beaucoup en lisibilié après ces changements 🙌 🔥 !!
Je pense qu'il y a encore quelque edge cases à gérer et on peut peut-être rajouter quelques tests supplémentaire, mais sinon c'est tout bon bravo :) 👏 !!

File: .eslintrc.json [hunk 4]
```diff
@@ -20,6 +20,8 @@
     "no-param-reassign": "off",
     "class-methods-use-this": "off",
     "no-nested-ternary": "off",
-    "no-use-before-define": "off"
+    "no-use-before-define": "off",
+    "react/jsx-props-no-spreading": "off",
+    "no-restricted-syntax": "off"
```
[the remark this answers is not in the document]
Comment (Michaelvilleneuve) [reply .eslintrc.json:25]: Je l'ai désactivé à cause de ça : 
```
for (const applicant of applicants.selectedApplicants) {
      // Using await here to avoid sending too many requests at the same time
      // eslint-disable-next-line no-await-in-loop
      await applicant.inviteBy(format, isDepartmentLevel);
}
```
La règle empêche de faire des boucles simple et veut privilégier un `.forEach` ou un map/reduce
Sauf que pour ce cas précis on voulait volontairement une boucle synchrone. 

Tu préfères que je comment cette règle uniquement dans le fichier en question ? ça permettra d'expliquer plus clairement je pense

[the remark this answers is not in the document]
Comment (aminedhobb) [reply .eslintrc.json:25]: Oui ça me parait mieux merci 🙏

File: app/javascript/lib/retrieveRelevantOrganisation.js [hunk 5]
```diff
@@ -5,19 +5,22 @@ import searchOrganisations from "../react/actions/searchOrganisations";
...
-  return retrieveThroughGeolocalisation(departmentNumber, applicantFullAddress);
+  return retrieveThroughGeolocalisation(departmentNumber, applicantFullAddress, options);
 };
 
-const retrieveThroughSearchTerms = async (departmentNumber, organisationSearchTerms) => {
+const retrieveThroughSearchTerms = async (departmentNumber, organisationSearchTerms, options = { raiseError: true }) => {
   const result = await searchOrganisations(departmentNumber, organisationSearchTerms);
   if (result.success && result.matching_organisations.length === 1) {
     return result.matching_organisations[0];
+  } if (options.raiseError === false) {
```
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/lib/retrieveRelevantOrganisation.js:22]: haha ah ouais bizarre... Et même si tu mets le `if` sur la ligne d'après (c'est déjà mieux je pense) ? 
Au pire on peut désactiver la règle 🤷‍♂️

File: app/javascript/react/components/ApplicantBatchActions.jsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  const deleteAll = () => {
+    toggle()
+    applicants.setApplicants(applicants.list.filter((applicant) => !applicant.selected))
+  }
+
+  const inviteBy = async (format) => {
+    toggle()
+    for (const applicant of applicants.selectedApplicants) {
```
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/react/components/ApplicantBatchActions.jsx:21]: Ok merci pour la réponse 👍

File: app/javascript/react/components/ApplicantBatchActions.jsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      <div style={{ marginRight: 20 }}>
+        <button type="button" className="btn btn-primary dropdown-toggle" onClick={toggle}>
+          Actions pour toute la sélection
+        </button>
+        <div className="dropdown-menu" id="batch-actions">
+          <button type="button" className="dropdown-item d-flex justify-content-between align-items-center" onClick={createAccounts}>
+            <span>Créer comptes</span>
+            <i className="fas fa-user" />
+          </button>
+          <button type="button" className="dropdown-item d-flex justify-content-between align-items-center" onClick={() => inviteBy("email")}>
```
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/react/components/ApplicantBatchActions.jsx:47]: En fait si tu vois la méthode `canBeInvitedBy` elle dépend seulement de la configuration au niveau de l'organisation et non pas de chaque `applicant` (ce model est un peu fourre-tout 😅 ) donc ça s'applique pour tous ici

File: app/javascript/react/pages/ApplicantsUpload.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/react/pages/ApplicantsUpload.jsx:135]: On peut après ça va peut-être un peu fastidieux de réimplémenter notre logique d'identification juste pour ce `find`. 
Peut-être que le fait de vider la liste à chaque upload suffit ? J'ai l'impression que c'est le comportement attendu

File: app/javascript/lib/retrieveRelevantOrganisation.js (hunk 1 above)
Comment (aminedhobb) [reply app/javascript/lib/retrieveRelevantOrganisation.js:23]: Oui bien sûr, normalement si tu uploades le fichier au niveau du département (= `Toutes les organisations`), on devrait te proposer cette modale vu qu'il n'y a pas de sectorisation de configurée sur RDVS ni de colonne correspondant au nom de l'organisation dans le fichier

File: app/javascript/react/models/Applicant.js (hunk 3 above)
Comment (aminedhobb) [reply app/javascript/react/models/Applicant.js:135]: Ouais je me suis fait la réflexion aussi mais à première vue je vois pas d'autres solutions non plus 🤷‍♂️ . Je pense que c'est acceptable en l'état.

File: app/javascript/react/pages/ApplicantsUpload.jsx [hunk 8]
```diff
@@ -338,4 +346,9 @@ export default function ApplicantsUpload({
       )}
     </div>
   );
-}
+})
+
+
+export default (props) => (
+  <ApplicantsUpload applicants={Applicants} {...props} />
```
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/react/pages/ApplicantsUpload.jsx:353]: Ok je comprends. Mais alors c'est peut-être la notation qui me perturbe, j'ai l'impression ici qu'on passe une classe alors qu'on passe une instance de cette classe.
Est-ce qu'on peut du coup instancier une variable `applicants` dans `/models/Applicants` qu'on exporte et qu'on réimporte dans `ApplicantsUpload` ?

File: app/javascript/react/models/Applicant.js [hunk 9]
```diff
@@ -27,6 +31,10 @@ export default class Applicant {
     Object.keys(attributes).forEach((key) => {
       formattedAttributes[key] = attributes[key]?.toString()?.trim();
     });
+    this._id = formattedAttributes.id;
+    this._createdAt = formattedAttributes.createdAt;
+    this._organisations = formattedAttributes.organisations || [];
```
[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/react/models/Applicant.js:36]: Yes carrément, j'ai créé une issue ici :) : https://github.com/betagouv/rdv-insertion/issues/1225

File: app/javascript/react/models/Applicant.js (hunk 3 above)
Comment (Michaelvilleneuve) [reply app/javascript/react/models/Applicant.js:135]: En fait techniquement ça ne devrait pas être à la "couche réseau" de lever des erreurs UI c'est pas trop sa responsabilité mais bon c'était clairement pratique pour gérer les erreurs sans avoir à gérer des states d'erreurs j'imagine

File: app/javascript/react/pages/ApplicantsUpload.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Michaelvilleneuve) [reply app/javascript/react/pages/ApplicantsUpload.jsx:135]: Mmh oui tu as raison 👍

File: app/javascript/react/pages/ApplicantsUpload.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (Michaelvilleneuve) [reply app/javascript/react/pages/ApplicantsUpload.jsx:354]: Ouai je vois ce que tu veux dire, je pense que c'est pour ça qu'en général avec mobx on nomme ce genre de dossiers non pas `models` mais `stores` ce qui donne plus la sensation d'importer un truc déjà instanciée 

Sinon sachant que c'est un export par défaut je peux juste l'importer avec comme nom `applicants` en minuscule, ça t'irait ?

[the remark this answers is not in the document]
Comment (aminedhobb) [reply app/javascript/react/pages/ApplicantsUpload.jsx:354]: Oui tu as raison. LGTM 👍

File: app/javascript/react/models/Applicant.js (hunk 3 above)
Comment (aminedhobb) [reply app/javascript/react/models/Applicant.js:135]: Oui c'est vrai on a pas architecturé comme il faudrait (on a pas trop voulu investir sur la partie react de l'app qui n'est pas amenée à s'étendre), on pourrait voir ça ultérieurement si nécessaire

**Review by aminedhobb — APPROVED** [review]
Merci pour les changements @Michaelvilleneuve 🙌 🙏 !
Let's go 🚀 !!
